File: stack/python/test_stack.py
import pytest

from .stack import Stack
from .item import Item


@pytest.fixture
def sample_stack():
    stack = Stack()
    for i in range(5):
        stack.push(Item(i+1))
    return stack


def test_can_creat_stack():
    non_sized_stack = Stack()
    sized_stack = Stack(10)

    assert non_sized_stack is not None
    assert non_sized_stack.size == 0
    assert non_sized_stack.length == 0

    assert sized_stack is not None
    assert sized_stack.size == 10
    assert sized_stack.length == 0


def test_can_push_on_empty_stack():
    stack = Stack()
    item = Item(10)

    pushed = stack.push(item)

    assert pushed is True
    assert stack.length == 1
    assert stack.top.item.value == 10


def test_can_push_on_non_empty_stack(sample_stack):
    item = Item(10)

    pushed = sample_stack.push(item)

    assert pushed is True
    assert sample_stack.length == 6
    assert sample_stack.top.item.value == 10


def test_cannot_push_on_full_stack(sample_stack):
    sample_stack.size = 5
    item = Item(10)

    pushed = sample_stack.push(item)

    assert pushed is False
    assert sample_stack.length == 5
    assert sample_stack.top.item.value == 5


def test_can_pop_on_empty_stack():
    stack = Stack()

    popped = stack.pop()

    assert(popped is None)
    assert(stack.length == 0)


def test_can_pop_on_non_empty_stack(sample_stack):
    popped = sample_stack.pop()

    assert popped is not None
    assert popped.value == 5
    assert sample_stack.length == 4
    assert sample_stack.top.item.value == 4


def test_can_peek_on_empty_stack():
    stack = Stack()

    peeked = stack.peek()

    assert peeked is None
    assert(stack.length == 0)


def test_can_peek_on_non_empty_stack(sample_stack):
    peeked = sample_stack.peek()

    assert peeked is not None
    assert sample_stack.length == 5
    assert sample_stack.top.item.value == 5

File: stack/c/test_stack.c
#include "stack.h"

#include <stdio.h>
#include <assert.h>


static lstack_t* creat_sample_stack(int size, int num_items) {
    lstack_t *stack = create_stack(size);
    for(int i=1; i <= num_items; i++) {
        push_on_stack(stack, create_item(i));
    }
    return stack;
}


void test_can_create_stack() {

    lstack_t *no_size_stack = create_stack(0);
    lstack_t *sized_stack = create_stack(10);

    assert(no_size_stack->size == 0);
    assert(no_size_stack->length == 0);
    assert(no_size_stack->top == NULL);

    assert(sized_stack->size == 10);
    assert(sized_stack->length == 0);
    assert(sized_stack->top == NULL);

    printf("PASS: %s\n", __func__);
}


void test_can_create_item() {
    item_t *item = create_item(10);

    assert(item != NULL);
    assert(item->value == 10);

    printf("PASS: %s\n", __func__);
}

void test_can_push_on_empty_stack() {
    lstack_t *stack = create_stack(0);
    item_t *item = create_item(10);

    int pushed = push_on_stack(stack, item);

    assert(pushed == 1);
    assert(stack->length == 1);
    assert(stack->top->item->value == 10);

    printf("PASS: %s\n", __func__);
}


void test_can_push_on_non_empty_stack() {
    lstack_t *stack = creat_sample_stack(0, 5);
    item_t *item = create_item(10);
    int initial_stack_length = stack->length;

    int pushed = push_on_stack(stack, item);

    assert(pushed == 1);
    assert(stack->length == 6);
    assert(stack->top->item->value == 10);

    printf("PASS: %s\n", __func__);
}


void test_cannot_push_on_full_stack() {
    lstack_t *stack = creat_sample_stack(5, 5);
    item_t *item = create_item(10);
    int initial_stack_length = stack->length;

    int pushed = push_on_stack(stack, item);

    assert(pushed == 0);
    assert(stack->length == 5);
    assert(stack->top->item->value == 5);

    printf("PASS: %s\n", __func__);
}


void test_can_pop_on_empty_stack() {
    lstack_t *stack = create_stack(0);

    item_t *popped = pop_on_stack(stack);

    assert(popped == NULL);
    assert(stack->length == 0);

    printf("PASS: %s\n", __func__);
}


void test_can_pop_on_non_empty_stack() {
    lstack_t *stack = creat_sample_stack(5, 5);

    item_t *popped = pop_on_stack(stack);

    assert(popped != NULL);
    assert(popped->value == 5);
    assert(stack->length == 4);
    assert(stack->top->item->value == 4);

    printf("PASS: %s\n", __func__);
}


void test_can_peek_on_empty_stack() {
    lstack_t *stack = create_stack(0);

    item_t *peeked = peek_on_stack(stack);

    assert(peeked == NULL);
    assert(stack->length == 0);

    printf("PASS: %s\n", __func__);
}


void test_can_peek_on_non_empty_stack() {
   lstack_t *stack = creat_sample_stack(5, 5);

    item_t *peeked = peek_on_stack(stack);

    assert(peeked != NULL);
    assert(stack->length == 5);
    assert(stack->top->item->value == 5);

    printf("PASS: %s\n", __func__);
}


int main() {
    test_can_create_stack();
    test_can_create_item();
    test_can_push_on_empty_stack();
    test_can_push_on_non_empty_stack();
    test_cannot_push_on_full_stack();
    test_can_pop_on_empty_stack();
    test_can_pop_on_non_empty_stack();
    test_can_peek_on_empty_stack();
    test_can_peek_on_non_empty_stack();
}